src/shim/hwctx.cpp

```diff
@@ -244,14 +244,31 @@ delete_ctx_on_device()
   fini_log_buf();
 }
 
+void
+hw_ctx::
+set_metadata(int num_cols, size_t size)
+{
+  m_metadata.magic_no = 0x434552544C4F4753;
+  m_metadata.major = 0;
+  m_metadata.minor = 1;
+  m_metadata.num_cols = num_cols;
+  for (int i = 0; i < num_cols; i++) {
+    m_metadata.col_offset[i] = size * i;
```

Comment:
There are 2 additional thoughts here:
1). the offset should be considered the metadata size at the offset 0x0, then the real data. By anytime, the metadata should not be corrupted by the data. Thus the first offset should be the sizeof(metadata) + i * size.
2). I'd think we are able to add device (absolute dma address) instead of offset, thus the firmware doesn't have to be recalculate the address. I will update the design.